### PR TITLE
Wave track affordance rewrites

### PR DESCRIPTION
**Pull request by vsverchinsky**
Resolves: #5076 

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: src/tracks/playabletrack/wavetrack/ui/WaveTrackAffordanceControls.h [hunk 1]
```diff
@@ -36,20 +38,22 @@ class AUDACITY_DLL_API WaveTrackAffordanceControls :
     public TextEditDelegate,
     public std::enable_shared_from_this<WaveTrackAffordanceControls>
 {
-    std::weak_ptr<WaveClip> mFocusClip;
+   using IntervalIterator = ChannelGroup::IntervalIterator<WaveTrack::Interval>;
```
Comment (Paul-Licameli) [inline src/tracks/playabletrack/wavetrack/ui/WaveTrackAffordanceControls.h:41]: Indentation

File: src/tracks/playabletrack/wavetrack/ui/WaveChannelView.cpp [hunk 2]
```diff
@@ -1705,7 +1705,10 @@ std::weak_ptr<WaveClip> WaveChannelView::GetSelectedClip()
 {
    if (auto affordance = std::dynamic_pointer_cast<WaveTrackAffordanceControls>(GetAffordanceControls()))
    {
-      return affordance->GetSelectedClip();
+      if(auto interval = *affordance->GetSelectedInterval())
+      {
+         return interval->GetClip(GetChannelIndex());
```
Comment (Paul-Licameli) [inline src/tracks/playabletrack/wavetrack/ui/WaveChannelView.cpp:1710]: I want only constant `0` as an argument of `GetClip` and later I will remove the argument.

But the outer `if` should justify `assert(GetChannelIndex() == 0);`

So write that assertion, then change the argument of `GetClip`.

Actually `GetChannelIndex()` still always returns 0, unlike `ReallyGetChannelIndex`

The problem of unifying those two functions (keeping the name of the first, and the behavior of the second) is another box on my checklist for later.

**Comment by Paul-Licameli** [comment]
I reviewed the rest of it, and have nothing more to comment about, except this:

You store `ChannelIterator` objects, and those have plain C-pointers to `GroupType` in them.  But I don't think there is anything in the contracts of these classes giving good enough lifetime guarantees against that becoming a dangling pointer.

I think nothing wrong will happen and I have no specific rewrite to insist that you do, but it bothers my mind a little.

**Comment by Paul-Licameli** [comment]
So, consider this round of review complete, with just one easy request for changes.

**Comment by vsverchinsky** [comment]
> I reviewed the rest of it, and have nothing more to comment about, except this:
> 
> You store `ChannelIterator` objects, and those have plain C-pointers to `GroupType` in them. But I don't think there is anything in the contracts of these classes giving good enough lifetime guarantees against that becoming a dangling pointer.
> 
> I think nothing wrong will happen and I have no specific rewrite to insist that you do, but it bothers my mind a little.

Alternatively I could use indices

**Comment by Paul-Licameli** [comment]
> > I reviewed the rest of it, and have nothing more to comment about, except this:
> > You store `ChannelIterator` objects, and those have plain C-pointers to `GroupType` in them. But I don't think there is anything in the contracts of these classes giving good enough lifetime guarantees against that becoming a dangling pointer.
> > I think nothing wrong will happen and I have no specific rewrite to insist that you do, but it bothers my mind a little.
> 
> Alternatively I could use indices

Maybe, but I will approve it now.

**Review by Paul-Licameli — APPROVED** [review]
I approve, conditional on successful CI

**Comment by chinakov** [comment]
Tested

**Pull request merged**